User-guide: Add automated cleaning feature user document (#242)
metal3-io/metal3-docs: docs/user-guide/src/capm3/automated_cleaning.md

# Automated Cleaning

Before reading this page, please see [Baremetal Operator Automated Cleaning](../bmo/automated_cleaning.md) page.

If you are using only Metal3 Baremetal Operator, you can skip this page and refer to Baremetal

> Operator automated cleaning [page](../bmo/automated_cleaning.md) instead.

Reviewer: I think these two paragraphs can be merged into one since both are referring/pointing out to same doc and on the same direction

Author: Would be nice, but not sure how. Because one is as prerequisite and the letter is for only BMO

For deployments following Cluster-api-provider-metal3 (CAPM3) workflow, automated cleaning can
be (recommended) configured via CAPM3 custom resources (CR).

There are two automated cleaning modes available  which can be set via `automatedCleaningMode` field of a
Metal3MachineTemplate `spec` or Metal3Machine `spec`.

- `metadata` to enable the cleaning
- `disabled` to disable the cleaning

> When enabled (`metadata`), automated cleaning kicks off when a node is in the first provisioning and on every deprovisioning.

Reviewer 2: ```suggestion
When enabled (`metadata`), automated cleaning kicks off when a node is provisioning for the first  and on every time it is deprovisioning.
```

There is no defaulting for `automatedCleaningMode` in Metal3MachineTemplate and Metal3Machine. If user doesn't set any mode,

> the field in the `spec` will be omitted. Unsetting `automatedCleaningMode` in the Metal3MachineTemplate will block the synchronization of the cleaning mode between the Metal3MachineTemplate and Metal3Machines.

Reviewer 3: Maybe add here that "this enables the selective operations described below" or "this allows the Metal3Machines to have different values if desired"?
It wasn't clear to me before reading about the selective operations if "blocking the synchronization" was a "configuration error" or something that may be desired in some situations. With this addition I think it would be clearer.

Author: good point. Added

## Bulk operations

CAPM3 controller ensures to replicate automated cleaning mode to all Metal3Machines from their referenced Metal3MachineTemplate.
For example, one controlplane and one worker Metal3Machines have `automatedCleaningMode` set to `disabled`, because it is set to `disabled` also in the template that they both are referencing.

> **Note**: CAPM3 controller replicates the cleaning mode from Metal3MachineTemplate to Metal3Machine only if `automatedCleaningMode` is not unset (empty). In other words, it synchronizes either `disabled` or `metadata` modes between Metal3MachineTemplate and Metal3Machines.

Reviewer 3: ```suggestion
**Note**: CAPM3 controller replicates the cleaning mode from Metal3MachineTemplate to Metal3Machine only if `automatedCleaningMode` is set (not empty). In other words, it synchronizes either `disabled` or `metadata` modes between Metal3MachineTemplate and Metal3Machines.
```

## Selective operations

Normally automated cleaning mode is replicated from Metal3MachineTemplate `spec` to its child Metal3Machines' `spec` and from Metal3Machines `spec` to BareMetalHost `spec` (if CAPM3 is used). However, sometimes you might want to have a different automated cleaning mode for one or more Metal3Machines than the others even though they belong to one Metal3MachineTemplate. For example, there is one worker and one controlplane Metal3Machine created from the same Metal3MachineTemplate, and we would like the automated cleaning to be enabled (`metadata`) for the worker while disabled (`disabled`) for the controlplane.

Here are the steps to achieve that:

> 1. Unset `automatedCleaningMode` in the Metal3MachineTemplate. Then CAPM3 controller unsets it for child Metal3Machines. Although it is unset in the Metal3Machine, BareMetalHosts will get their default automated cleaning mode `metadata`. As we mentioned earlier, CAPM3 controller replicates cleaning mode from Metal3MachineTemplate to Metal3Machine ONLY when it is either `metadata` or `disabled`. As such, to block synchronization between Metal3MachineTemplate and Metal3Machine, unsetting the cleaning mode in the Metal3MachineTemplate is enough.

Reviewer 4: ```suggestion
1. Unset `automatedCleaningMode` in the Metal3MachineTemplate. Then CAPM3 controller unsets it for referencing Metal3Machines. Although it is unset in the Metal3Machine, BareMetalHosts will get their default automated cleaning mode `metadata`. As we mentioned earlier, CAPM3 controller replicates cleaning mode from Metal3MachineTemplate to Metal3Machine ONLY when it is either `metadata` or `disabled`. As such, to block synchronization between Metal3MachineTemplate and Metal3Machine, unsetting the cleaning mode in the Metal3MachineTemplate is enough.
```

> 1. Set `automatedCleaningMode` to `disabled` on the worker Metal3Machine `spec` and to `metadata` on the controlplane Metal3Machine `spec`. Since we don't have any mode set on the Metal3MachineTemplate, Metal3Machines can have different automated cleaning modes set even if they reference the same Metal3MachineTemplate. CAPM3 controller copies cleaning modes from Metal3Machines to their corresponding BareMetalHosts. As such, we end up with two nodes having different cleaning modes regardless of the fact that they belong to the same Metal3MachineTemplate.

Reviewer 4: ```suggestion
1. Set `automatedCleaningMode` to `disabled` on the worker Metal3Machine `spec` and to `metadata` on the controlplane Metal3Machine `spec`. Since we don't have any mode set on the Metal3MachineTemplate, Metal3Machines can have different automated cleaning modes set even if they reference the same Metal3MachineTemplate. CAPM3 controller copies cleaning modes from Metal3Machines to their corresponding BareMetalHosts. As such, we end up with two nodes having different cleaning modes regardless of the fact that they referencing the same Metal3MachineTemplate.
```
